enhancements/installer/agent-based-installation-in-hive.md

```diff
@@ -0,0 +1,539 @@
+
+The first implementation of controllers that provide a CRD-based API will be
+added to the existing
+[assisted-service](https://github.com/openshift/assisted-service) code in a way
+that allows the existing REST API to continue existing and running
+side-by-side, including its database. An end user, whether human or automation,
+will only need to interact with the CRDs; the database and REST API will be
+implementation details that can be removed in the future. While continuing to
+run a separate database and continuing to utilize parts of the existing REST
+API are not ideal for an in-cluster service, this approach is the only
+practical way to deliver a working solution in a short time frame.
+
+While making a separate version that is Kubernetes-native with no SQL DB is on
+the table for the long term, it is not feasible to implement and maintain such
+a different pattern in the near-term.
+
+#### Kubernetes-native APIs for Agent Based Cluster Provisioning
+
+The Assisted Installer has a [REST
+API](https://generator.swagger.io/?url=https://raw.githubusercontent.com/openshift/assisted-service/master/swagger.yaml)
+that is available at cloud.redhat.com. The service is implemented with a
+traditional relational database. In order to integrate with OpenShift
+infrastructure management tooling, it is necessary to additionally expose its
+capabilities as Kubernetes-native APIs. They will be implemented as Custom
+Resource Definitions and controllers with the Assisted Service's capabilities.
+
+Hive's ClusterDeployment resource will be extended to serve as the cluster
+definition for agent-based installations. Additional CRDs described below will
+be used by other controllers and actors to perform agent-based provisioning.
+
+For the first deliverable, a local deployment of the backing service will
+include new controllers that will operate to achieve the desired state as
+declared in CRDs. Thus the service will expose both REST and Kubernetes-native
+APIs, and both API layers will call the same set of internal methods. The local
+service will include a database; sqlite if possible, or else postgresql.
+
+    --------------------------
+    |  REST API  |  k8s API  |
+    --------------------------
+    | Assisted Service logic |
+    --------------------------
+        |               |
+        v               v
+      SQL DB        File System
+                      storage
+
+For hub clusters that are used for multi-cluster creation and management,
+persistent storage will be a requirement for the database and file system
+storage.
+
+Agents that are running on hardware and waiting for installation-related
+instructions will continue to communicate with the assisted service using the
+existing REST API. In the future that may transition to using a CRD directly.
+In the meantime, status will be propagated to the appropriate k8s resource when
+an agent reports relevant information.
+
+**InstallEnv**
```

Comment:
Is this resource namespaced or not?

How do we intend to deal with RBAC and multi-tenancy, are we worried about someone taking control of Agents just by applying an arbitrary label in their ClusterDeployment?

Reply:
Yes it's namespaced, and I think that's what we use for multi-tenancy; label-based matching has to happen within a namespace.

Reply:
This is now stated explicitly in a couple of places, including the `#### Namespaces and Multi-tenancy` section.